shinysdr/plugins/wspr/demodulator.py

```diff
@@ -0,0 +1,196 @@
+from zope.interface import implementer
+
+from shinysdr.values import ExportedState, exported_value
+from shinysdr.math import dB
+from shinysdr.interfaces import IDemodulator, BandShape
+from shinysdr.signals import SignalType
+
+from .blocks import WAVIntervalSink, WSPRFilter
+from .interfaces import IWAVIntervalListener
+from .telemetry import WSPRSpot
+
+
+@implementer(IDemodulator)
+class WSPRDemodulator(gr.hier_block2, ExportedState):
+    '''Decode WSPR (Weak Signal Propagation Reporter).
+    
+    Requires `wsprd` to be installed, which is available as part of WSJT-X:
+
+    https://physics.princeton.edu/pulsar/k1jt/wsjtx.html
+    '''
+
+    # filename of wav file currently being recorded, or None if no recording is
+    # happening now.
+    __recording_filename = None
+
+    # wsprd requires wav files at 12kHz sample rates. The WSPR band is 200 Hz
+    # wide, centered on 1500 Hz in the recording. Need to confirm this, but I
+    # suspect it doesn't try to decode much outside that range.
+    #
+    # Our job is to make those recordings, starting on even minutes. Then they
+    # are passed to `wsprd` to do the decoding.
+
+    __demod_rate = 12000
+    __transition_width = 100
+
+    # transmission interval, in seconds. 2 minutes for WSPR. Might be adapted
+    # to other JT modes which use other intervals.
+    __interval = 120
+    __duration = __interval - 5
+
+    # TODO: look for this in $PATH
+    __wsprd = '/usr/local/bin/wsprd'
+
+    # TODO: get this from configuration?
+    __recording_dir = '/tmp'
+
+    def __init__(
+        self,
+        mode='WSPR',
+        input_rate=0,
+        context=None,
+
+        _WAVIntervalSink=WAVIntervalSink,
+    ):
+        assert input_rate > 0
+        gr.hier_block2.__init__(
+            self, 'WSPR Demodulator',
+            gr.io_signature(1, 1, gr.sizeof_gr_complex),
+            gr.io_signature(1, 1, gr.sizeof_float))
+        self.__context = context
+
+        # Step 1: filter down to 12kHz float samples, with the WSPR signals
+        # between 1400 and 1600 Hz.
+
+        wspr_filter = WSPRFilter(input_rate)
+
+        self.connect(self, wspr_filter)
+
+        # Step 2: audio out for monitoring.
+
+        self.connect(
+            wspr_filter,
+
+            analog.agc2_ff(
+                reference=dB(-10),
+                attack_rate=8e-1,
+                decay_rate=8e-1),
+
+            self)
+
+        # Step 3: output to wav file for decoding by wsprd.
+
+        listener = WAVIntervalListener(self.__recording_dir, context)
+
+        self.__wav_sink = _WAVIntervalSink(
+            interval=self.__interval,
+            duration=self.__duration,
+            listener=listener,
+            sample_rate=self.__demod_rate,
+        )
+
+        # would be cool to not have side effects in __init__, though there
+        # doesn't seem to be a way around it with the current demodulator
+        # interface.
+        self.__wav_sink.start_running()
```

Comment:
If no, what do you think about separating `IDemodulator` from `gr.hier_block2`?

* The current design seems to be a violation of the "composition over inheritance" principle
* Mixing things that run in the GNU Radio thread with things called from the reactor is very tricky -- easy to lose track of what's called from where and end up blocking the reactor, or calling non-thread-safe GNU Radio things in a bad way.
* There doesn't seem to be anything in `IDemodulator` actually specifying the GNU Radio block part of the interface.